### PR TITLE
Create a Feature Branch Approval Policy

### PR DESCRIPTION
**Pull request by mattcaswell**
Defines a policy for how new feature branches should be requested and approved. We defer the management of feature branches to other future policies.

Fixes openssl/project#423

### CODE REVIEW

**Comment by mattcaswell** [comment]
See openssl/project#423

File: policies/feature-branch-approval-policy.md
Comment (t8m) [inline policies/feature-branch-approval-policy.md:30]: My suggestion:
```suggestion
- The name of the feature branch requested in the form of `feature/<name>`
```

This way we will use the same pattern as for the already existing `feature/quic-server` branch.

**Comment by t8m** [comment]
LGTM

**Comment by hlandau** [comment]
Seems reasonable to me. I think it needs to discuss rebasing a bit, I don't think that can be deferred.

The protocol I proposed when creating the QUIC server feature branch was that rebasing can be done immediately after getting an informal OK from another committer.

However, we need to determine how to review nontrivial rebases where changes had to be made during the rebase process.

**Comment by hlandau** [comment]
Hmm, though. I'm not entirely 100% about saying a feature branch can be merged without further review. I think it should be done as part of a PR, even if we agree that we don't actually need to go and review all the changes in that PR because the contents have already been reviewed individually. This will be a PR between two branches in the main repository, rather than the main repository and someone's fork. It can then be given a simple procedural approval in the usual manner, basically just glancing at it to make sure there's nothing off about the changes (e.g. if a PR to merge a QUIC feature branch had changes to crypto/aes, there's obviously something weird there). It would be similar to the current rather rudimentary review and approval of fuzz-corpora changes. Also ensures it is confirmed that it has been determined that it is OK to merge a feature branch at that particular time with commitment to a release.

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:7]:
```suggestion
implemented and reviewed within a single PR there is a risk that only a
partial implementation of the feature will be merged to the master branch
by the time a feature freeze occurs for the next release.
```

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:3]:
```suggestion
The [Time-based Release Policy] defines a regular twice-a-year schedule for new
```

Comment (arapov) [reply policies/feature-branch-approval-policy.md:3]: It feels like it should have hyphens...

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:20]:
```suggestion
PRs targeting it). Such a feature branch would typically only be created where
```

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:28]:
```suggestion
anyone. The request should specify:
```

Comment (arapov) [reply policies/feature-branch-approval-policy.md:28]: Simplify?

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:47]:
```suggestion
all commits may then be eligible for merging into the master branch, subject to
final review.
```

Comment (arapov) [reply policies/feature-branch-approval-policy.md:47]: Merging without review sounds dangerous and reckless, although I understand that everything merged into the feature branch has already been reviewed and did not show issues in CI.

Nonetheless, we should conduct an extra review. It might be less thorough, but a final check sounds reasonable.

Wdyt?

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:34]: The requester will have difficulty determining this as it depends on our capacity to review. Should it be framed as an estimation of when the work on the feature is expected to be done? Ultimately, it is up to us which release we will accept.

File: policies/feature-branch-approval-policy.md
Comment (arapov) [inline policies/feature-branch-approval-policy.md:30]: Should we put here other requirements?
- Feature detailed description;
- The purpose of the feature [and how it aligns with the project's goals];
- Expected scope and impact of this feature. Any changes to existing functionality?
- Potential benefits or drawbacks;
- Dependencies;
- Plan to test? Unit tests, integration tests, and any manual testing procedures.
- Documentation;
- Responsibility for maintaining this feature once it's integrated;

File: policies/feature-branch-approval-policy.md
Comment (t8m) [reply policies/feature-branch-approval-policy.md:7]: maybe s/only a partial/an incomplete or incorrect/

File: policies/feature-branch-approval-policy.md
Comment (t8m) [reply policies/feature-branch-approval-policy.md:30]: This sounds like too much bureaucratic boilerplate. Also there will be an accompanying openssl/openssl feature issue linked where most of this should be described already. I.e., the openssl/project feature branch request should happen only after the feature is already discussed in the openssl/openssl feature issue and there is a consensus that doing such feature in a single PR is not reasonable.

> Responsibility for maintaining this feature once it's integrated;

That is an interesting point - I do not think we should ever accept a feature that we are not willing to maintain as the team. BTW - how we can hold the contributors to be responsible for this promised responsibility if they simply later disappear?

File: policies/feature-branch-approval-policy.md
Comment (mattcaswell) [reply policies/feature-branch-approval-policy.md]: I've added some words about this

File: policies/feature-branch-approval-policy.md
Comment (mattcaswell) [reply policies/feature-branch-approval-policy.md]: I've not made a change here, as per @t8m's comments

**Comment by mattcaswell** [comment]
Update pushed. Please take another look

File: policies/feature-branch-approval-policy.md
Comment (t8m) [inline policies/feature-branch-approval-policy.md:16]: Typo: `in an`

**Comment by mattcaswell** [comment]
Updated again to resolve @t8m's nit.

File: policies/feature-branch-approval-policy.md
Comment (mattcaswell) [reply policies/feature-branch-approval-policy.md]: @arapov - my suggestion is that we move some of these things into the feature request issue template:

https://github.com/openssl/openssl/blob/master/.github/ISSUE_TEMPLATE/feature_request.md?plain=1

**Review by t8m — APPROVED** [review]

**Comment by Fitnesskrokodyl** [comment]
We will have to discuss this next week - with Anton.

**Comment by mattcaswell** [comment]
Policy proposal has been announced on openssl-project

**Comment by rsbeckerca** [comment]
Does this policy permit community supported platforms? I would not want to see platform-specific features, but there are situations where code has to be added through a PR to support a platform's ongoing participation in OpenSSL.

**Comment by mattcaswell** [comment]
> Does this policy permit community supported platforms? I would not want to see platform-specific features, but there are situations where code has to be added through a PR to support a platform's ongoing participation in OpenSSL.

The policy says nothing about what type of features can be added through feature branches, and whether or not they are platform-specific. In general most platform-specific changes tend to be relatively small tweaks here and there which can usually be contained within a single PR (in which case this policy is not relevant). If however a larger change was necessary for something platform specific, and that change was too large for a single PR, then this policy could be used for it.

**Comment by eliskaman** [comment]
@mattcaswell We have to wait another week to vote. Let´s update next week.

**Comment by fwh-dc** [comment]
What is the policy for reviews of pull request to a feature branch?

Would a pull request with failing tests be acceptable in some scenarios for example?

**Comment by t8m** [comment]
There is no policy yet. However when the feature branches were discussed the general consensus at that time was that for the PRs on feature branches the same criteria should apply as for the master branch.

**Comment by mattcaswell** [comment]
Topic: Accept the feature branch approval policy as of d62052487a0656d80b4a5972625bedf06e20c94f
Proposed by: Matt Caswell
Issue link: https://github.com/openssl/general-policies/pull/62
Public: yes
Opened: 2024-02-20
Closed:
Accepted:  yes/no  (for: 0, against: 0, abstained: 0, not voted: 0)

  Anton      [  ]
  Kurt       [  ]
  Matt       [+1]
  Richard    [  ]
  Tim        [  ]

**Comment by mattcaswell** [comment]
I have called this to vote on openssl-project

**Comment by levitte** [comment]
Vote: +1

**Comment by t-j-h** [comment]
Vote: +1

**Comment by arapov** [comment]
+1

**Comment by mattcaswell** [comment]
@kroeckx - please could you vote?

**Comment by mattcaswell** [comment]
I have closed this vote and pushed this PR:

````
Topic: Accept the feature branch approval policy as of d62052487a0656d80b4a5972625bedf06e20c94f
Proposed by: Matt Caswell
Issue link: https://github.com/openssl/general-policies/pull/62
Public: yes
Opened: 2024-02-20
Closed: 2024-02-29
Accepted:  yes  (for: 4, against: 0, abstained: 0, not voted: 1)

  Anton      [+1]
  Kurt       [  ]
  Matt       [+1]
  Richard    [+1]
  Tim        [+1]
````

**Pull request closed without merging**
